### PR TITLE
Add support for Release CRs

### PR DESCRIPTION
**Pull request by nprokopic**
Towards https://github.com/giantswarm/roadmap/issues/3466

### What this PR does / why we need it

This PR adds support for using Release CRs and getting k8s, OS and app version, catalog and dependencies from there.

Will add more details here, for now see https://github.com/giantswarm/rfc/pull/96 for more.

### Checklist

- [x] Updated CHANGELOG.md.

### CODE REVIEW

File: helm/cluster-aws/values.schema.json [hunk 1]
```diff
@@ -488,6 +488,9 @@
                             "infrastructureMachineTemplateSpecTemplateName": "controlplane-awsmachinetemplate-spec"
                         }
                     },
+                    "osImage": {
```
Comment (nprokopic) [inline helm/cluster-aws/values.schema.json:491]: I will actually remove this, as variant will be also in the Release CR.

**Comment by nprokopic** [comment]
Will rebase and update PR.

**Review by mnitchev — APPROVED** [review]

File: helm/cluster-aws/templates/_helpers.tpl [hunk 2]
```diff
@@ -65,14 +65,13 @@ ami:
   id: {{ . | quote }}
 {{- else -}}
 ami: {}
-imageLookupBaseOS: "flatcar-stable"
-{{- $suffix := .Values.global.providerSpecific.osImageVariant }}
-{{- if $suffix }}
-  {{- $suffix = printf "-%s-gs" $suffix }}
+{{- $suffix := include "cluster.component.flatcar.variant" $ }}
+imageLookupBaseOS: "{{ include "cluster.component.flatcar.version" $ }}"
+{{- if and $suffix (ne $suffix "N/A") }}
```
Comment (AndiDog) [inline helm/cluster-aws/templates/_helpers.tpl:70]: Can we please add `{{/* comments */}}` to explain what's going on here?

File: helm/cluster-aws/templates/aws-ebs-csi-driver-helmrelease.yaml [hunk 3]
```diff
@@ -33,9 +33,9 @@ spec:
   chart:
     spec:
       chart: aws-ebs-csi-driver-app
-      # used by renovate
-      # repo: giantswarm/aws-ebs-csi-driver-app
-      version: 2.30.1
+      {{- $_ := set $ "appName" "aws-ebs-csi-driver" }}
+      {{- $appVersion := include "cluster.app.version" $ }}
```
Comment (AndiDog) [inline helm/cluster-aws/templates/aws-ebs-csi-driver-helmrelease.yaml:37]: Also possible: `mustMergeOverwrite (mustDeepCopy $) (dict "appName" "myapp")` (not tested)

**Review by AndiDog — COMMENTED** [review]
Mostly ready to go, just minor stuff

File: helm/cluster-aws/templates/aws-ebs-csi-driver-helmrelease.yaml (hunk 3 above)
Comment (nprokopic) [reply helm/cluster-aws/templates/aws-ebs-csi-driver-helmrelease.yaml:37]: You mean to pass in a new dict (deep copy of `$` + needed value) so we don't modify `$`?

**Comment by nprokopic** [comment]
let's see if this works 🙈 


/run cluster-test-suites

**Comment by nprokopic** [comment]
From the [e2e tests](https://tekton.ci.giantswarm.io/#/namespaces/tekton-pipelines/pipelineruns/pr-cluster-aws-637-cluster-test-suitesvvlfz?pipelineTask=run-tests&taskRunName=pr-cluster-aws-637-cluster-test-suitesvvlfz-run-tests-2&step=run-tests) ATM 🥹 ❤️ 

```
  {"level":"info","ts":"2024-06-12T16:13:37Z","msg":"Cluster App is supported by Releases"}
  {"level":"info","ts":"2024-06-12T16:13:38Z","msg":"Release name: 'aws-25.0.1-t.86x803cgg3'"}
```

@AverageMarcus did a wonderful job!

**Comment by nprokopic** [comment]
/run cluster-test-suites TARGET_SUITES=./providers/capa/upgrade

**Comment by nprokopic** [comment]
Re-running tests with the new [v25.0.0-alpha.2](https://github.com/giantswarm/releases/releases/tag/aws%2Fv25.0.0-alpha.2) release (tests picks up the latest one)

/run cluster-test-suites

**Comment by nprokopic** [comment]
Re-running tests with the new [v25.0.0-alpha.3](https://github.com/giantswarm/releases/releases/tag/aws%2Fv25.0.0-alpha.3) release

/run cluster-test-suites

**Comment by nprokopic** [comment]
Upgrade fails with

```
  {"level":"info","ts":"2024-06-13T14:09:14Z","msg":"deployment kube-system/cert-exporter-deployment has 0/1 replicas available"}
  {"level":"info","ts":"2024-06-13T14:09:25Z","msg":"deployment kube-system/cert-exporter-deployment has 0/1 replicas available"}
  {"level":"info","ts":"2024-06-13T14:09:36Z","msg":"deployment kube-system/cert-exporter-deployment has 0/1 replicas available"}
  [FAILED] in [It] - /app/internal/common/basic.go:80 @ 06/13/24 14:09:39.345
• [FAILED] [900.001 seconds]
Basic upgrade test basic [It] has all its Deployments Ready (means all replicas are running)
/app/internal/common/basic.go:76

  [FAILED] Timed out after 900.001s.
  Expected success, but got an error:
      <*errors.errorString | 0xc000abd1f0>: 
      deployment kube-system/cert-exporter-deployment has 0/1 replicas available
      {
          s: "deployment kube-system/cert-exporter-deployment has 0/1 replicas available",
      }
  In [It] at: /app/internal/common/basic.go:80
```

**Comment by nprokopic** [comment]
/run cluster-test-suites TARGET_SUITES=./providers/capa/upgrade

**Comment by nprokopic** [comment]
debugging failures


/run cluster-test-suites TARGET_SUITES=./providers/capa/upgrade

**Comment by T-Kukawka** [comment]
what is the status here @nprokopic ?

File: helm/cluster-aws/templates/_helpers.tpl (hunk 2 above)
Comment (nprokopic) [reply helm/cluster-aws/templates/_helpers.tpl:75]: Added comments.

**Review by fiunchinho — CHANGES_REQUESTED** [review]
I don't think this is the right approach. I wrote some of my thoughts here https://github.com/giantswarm/rfc/pull/96#issuecomment-2178934882

**Comment by nprokopic** [comment]
Upgrade test expectedly failed with (issue with launch template updates):

```
Basic upgrade test basic [It] has all the worker nodes running
/app/internal/common/basic.go:63

  [FAILED] Timed out after 900.000s.
  Expected success, but got an error:
      <*errors.errorString | 0xc000d21850>: 
      unexpected number of nodes
      {
          s: "unexpected number of nodes",
      }
  In [It] at: /app/internal/common/basic.go:74 @ 06/20/24 13:51:52.738
```

CAPA Cilium ENI mode suite failed with:

```
  {"level":"info","ts":"2024-06-20T13:29:09Z","msg":"App status for 't-jp54o5zwvzjgt8io3c-kyverno-policy-operator' is not yet as expected: expectedStatus='deployed' actualStatus='pending-install' (reason: 'Initial install underway')"}

```

This has been happening randomly lately with apps from the security bundle 😞

File: helm/cluster-aws/templates/_helpers.tpl [hunk 4]
```diff
@@ -21,6 +21,7 @@ Common labels
 {{- include "labels.selector" $ }}
 helm.sh/chart: {{ include "chart" . | quote }}
 application.giantswarm.io/team: {{ index .Chart.Annotations "application.giantswarm.io/team" | quote }}
+release.giantswarm.io/version: {{ .Values.global.release.version | trimPrefix "v" | quote }}
```
Comment (fiunchinho) [inline helm/cluster-aws/templates/_helpers.tpl:24]: This common labels will get applied to the `cluster-aws` CAPI manifests. I don't think we need the release label in those manifests, as they are baked in the `cluster-aws` release. So I'd remove the label.

Comment (nprokopic) [reply helm/cluster-aws/templates/_helpers.tpl:24]: > This common labels will get applied to the cluster-aws CAPI manifests.

That was the intention, so it's easier to see from which release some resources are, as multiple releases can have the same cluster-aws version. e.g. you can easily get AWSCluster, AWSMachinePool, etc from all clusters for a specific release.

**Comment by nprokopic** [comment]
After having a conversation in Slack we agreed to continue with this.

**Comment by nprokopic** [comment]
Adding `skip/ci` label, as test failures are known and expected.

File: helm/cluster-aws/templates/_helpers.tpl (hunk 4 above)
Comment (fiunchinho) [reply helm/cluster-aws/templates/_helpers.tpl:24]: Oh right, I didn't think of that. It makes sense now.

**Pull request merged**